README.md
# sped-ibpt

Este repositório faz parte do framework [NFePHP](http://www.nfephp.org), como um auxilio na busca dos impostos incidentes sobre um determinado produto, utilizando o recurso gratuito oferecido pelo [IBPT](https://ibpt.com.br/).

*A Lei do Imposto na Nota (Lei nº 12.741/12, de 8 de dezembro de 2012) nasceu com o intuito de informar ao cidadão o quanto representa a parcela dos tributos que paga a cada compra realizada.*

*Assim, todo estabelecimento que efetuar vendas diretamente ao consumidor final **está obrigado** a incluir nos documentos fiscais ou equivalentes os impostos pagos, valores aproximados e percentuais.*

*Como consumidores finais incluem-se as pessoas físicas ou jurídicas que adquirem produtos ou serviços, por exemplo, para consumo próprio, materiais de uso ou consumo e ativo imobilizado.*

*As Microempresas e Empresas de Pequeno Porte optantes do Simples Nacional podem informar apenas a alíquota a que se encontram sujeitas nos termos do referido regime. Além disso, devem somar eventual incidência tributária anterior (IPI, substituição tributária, por exemplo).*

[![Join the chat at https://gitter.im/nfephp-org/sped-ibpt](https://badges.gitter.im/nfephp-org/sped-ibpt.svg)](https://gitter.im/nfephp-org/sped-ibpt?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

Esta é uma API simples para permitir o acesso ao recursos providos pelos serviços RestFul do [IBPT "Instituto Brasileiro de Planejamento e Tributação"](https://deolhonoimposto.ibpt.org.br/).

Para saber mais consulte a [documentação do IBPT](https://deolhonoimposto.ibpt.org.br/).

[![Actions Status][ico-workflow]][link-actions]

[![Latest Version on Packagist][ico-version]][link-packagist]
[![License][ico-license]][link-packagist]
[![Total Downloads][ico-downloads]][link-downloads]

[![Issues][ico-issues]][link-issues]
[![Forks][ico-forks]][link-forks]
[![Stars][ico-stars]][link-stars]

## Esclarecimentos

1. Os cupons (mod.65, ECF e SAT) e notas fiscais (mod.55) referentes à venda de mercadoria e serviços **devem informar o valor aproximado dos tributos federais, estaduais e municipais**, cuja incidência influencia na formação do preço final;

2. A apuração do valor dos impostos deverá ser feita em 3 campos (um por ente) com a soma das cargas tributárias aproximadas que incidem sobre cada mercadoria ou serviço, separadamente, inclusive nas hipóteses de regimes jurídicos tributários diferenciados dos respectivos fabricantes, varejistas e prestadores de serviços, quando couber.

3. Alternativemente, no caso de lojas físicas, as informações podem estar em painel afixado em local visível do estabelecimento comercial. Elas serão em percentual sobre o preço a ser pago, quando se tratar de tributo com alíquota *"ad valorem"* (sobre valor) ou em valores monetários (no caso de alíquota específica).

> NOTA: Para maiores detalhes vide [NT2013.003](https://www.nfe.fazenda.gov.br/portal/exibirArquivo.aspx?conteudo=RrvyORm641k=)

## Pré Requisitos

Antes de poder utilizar esta classe é necessário que você obtenha um TOKEN de acesso cadastrando a empresa no IBPT [página de Cadastro](https://deolhonoimposto.ibpt.org.br/Usuario/CriarConta)

- PHP >= 7.0
- php-curl
- php-json
- php-openssl

## Intalação

```
composer require nfephp-org/sped-ibpt
```

Ou adicione ao seu composer.json:
```
{
    "require": {
        "nfephp-org/sped-ibpt": "^2.0"
    }
}
```

# Métodos

## productTaxes
Este método consulta o webservice do IBPT e solicita os dados referentes aos impostos do produto solicitado.
Sendo:
```php

use NFePHP\Ibpt\Ibpt;

$token = "<indique seu token>"; //OBRIGATÓRIO
$cnpj = "<indique seu CNPJ>"; //OBRIGATÓRIO

$ncm = "60063210"; //OBRIGATÓRIO coloque o NCM do produto
$uf = 'SP'; //OBRIGATÓRIO coloque o estado que deseja saber os dados
$extarif = 0; //OBRIGATÓRIO indique o numero da exceção tarifaria, se existir ou deixe como zero
$codigoInterno = ''; //(OPCIONAL) indique o codigo interno do produto 
$descricao = 'Tecido';//OBRIGATÓRIO
$unidadeMedida = 'kg'; //OBRIGATÓRIO
$valor = '60.00'; //OBRIGATÓRIO
$gtin = 'SEM GTIN'; //OBRIGATÓRIO

//instancia a classe 
$ibpt = new Ibpt($cnpj, $token);

$resp = $ibpt->productTaxes(
    $uf,
    $ncm,
    $extarif,
    $descricao,
    $unidadeMedida,
    $valor,
    $gtin,
    $codigoInterno
);
```
Em caso de SUCESSO e com a localização do Produto solicitado irá retornar:
```php
stdClass Object
(
    [Codigo] => 60063210
    [UF] => SP
    [EX] => 0
    [Descricao] => Tecidos de malha de fibras sinteticas, tintos, de náilon ou de outras poliamidas
    [Nacional] => 13.45
    [Estadual] => 18
    [Importado] => 36.08
    [Municipal] => 0
    [Tipo] => 0
    [VigenciaInicio] => 01/04/2018
    [VigenciaFim] => 30/06/2018
    [Chave] => F3W1D7
    [Versao] => 18.1.B
    [Fonte] => IBPT/empresometro.com.br
    [Valor] => 60
    [ValorTributoNacional] => 8.07
    [ValorTributoEstadual] => 10.8
    [ValorTributoImportado] => 21.65
    [ValorTributoMunicipal] => 0
)
```
Em caso de não encontrar o produto pelo NCM, ou qualquer outro erro na comunicação, retornará algo como:
```php
stdClass Object
(
    [Codigo] => 
    [UF] => 
    [EX] => 0
    [Descricao] => 
    [Nacional] => 0
    [Estadual] => 0
    [Importado] => 0
    [Municipal] => 0
    [Tipo] => 
    [VigenciaInicio] => 
    [VigenciaFim] => 
    [Chave] => 
    [Versao] => 
    [Fonte] => 
    [Valor] => 60
    [ValorTributoNacional] => 0
    [ValorTributoEstadual] => 0
    [ValorTributoImportado] => 0
    [ValorTributoMunicipal] => 0
)
```

## serviceTaxes
Este método consulta o webservice do IBPT e solicita os dados referentes aos impostos do serviço solicitado.
Sendo:
```php

use NFePHP\Ibpt\Ibpt;

$token = "<indique seu token>"; //OBRIGATÓRIO
$cnpj = "<indique seu CNPJ>"; //OBRIGATÓRIO

$codigo = '0107';  //OBRIGATÓRIO numero LV116 ou NBM
$uf = 'SP'; // //OBRIGATÓRIO
$descricao = 'Suporte técnico em informática';  //OBRIGATÓRIO
$unidadeMedida = 'un';  //OBRIGATÓRIO
$valor = '500.00';  //OBRIGATÓRIO

//instancia a classe 
$ibpt = new Ibpt($cnpj, $token);

$resp = $ibpt->serviceTaxes(
    $uf,
    $codigo,
    $descricao,
    $unidadeMedida,
    $valor
);
```
Em caso de SUCESSO e com a localização do Serviço solicitado irá retornar:
```php
stdClass Object
(
    [Codigo] => 0107
    [UF] => SP
    [Descricao] => Suporte técnico em informática, inclusive instalação, configuração e manutenção de programas de computação e bancos de dados.
    [Tipo] => 2
    [Nacional] => 13.45
    [Estadual] => 0
    [Municipal] => 2.7
    [Importado] => 15.45
    [VigenciaInicio] => 01/04/2018
    [VigenciaFim] => 30/06/2018
    [Chave] => F3W1D7
    [Versao] => 18.1.B
    [Fonte] => IBPT/empresometro.com.br
    [Valor] => 500
    [ValorTributoNacional] => 67.25
    [ValorTributoEstadual] => 0
    [ValorTributoImportado] => 77.25
    [ValorTributoMunicipal] => 13.5
)
```
Em caso de não encontrar o produto pelo NCM, ou qualquer outro erro na comunicação, retornará algo como:
```php
stdClass Object
(
    [Codigo] => 
    [UF] => 
    [Descricao] => 
    [Tipo] => 
    [Nacional] => 0
    [Estadual] => 0
    [Municipal] => 0
    [Importado] => 0
    [VigenciaInicio] => 
    [VigenciaFim] => 
    [Chave] => 
    [Versao] => 
    [Fonte] => 
    [Valor] => 500
    [ValorTributoNacional] => 0
    [ValorTributoEstadual] => 0
    [ValorTributoImportado] => 0
    [ValorTributoMunicipal] => 0
)
```


## Credits
- Roberto L. Machado (owner)

## Change log

Acompanhe o [CHANGELOG](CHANGELOG.md) para maiores informações sobre as alterações recentes.

## Contributing

Para contribuir por favor observe o [CONTRIBUTING](CONTRIBUTING.md) e o  [Código de Conduta](CONDUCT.md) parea detalhes.

## Security

Caso você encontre algum problema relativo a segurança, por favor envie um email diretamente aos mantenedores do pacote ao invés de abrir um ISSUE.

## License

Este pacote está diponibilizado sob LGPLv3 ou MIT License (MIT). Leia  [Arquivo de Licença](LICENSE.md) para maiores informações.

[ico-workflow]: https://github.com/nfephp-org/sped-ibpt/actions/workflows/ci.yml/badge.svg?style=flat-square
[ico-stars]: https://img.shields.io/github/stars/nfephp-org/sped-ibpt.svg?style=flat-square
[ico-forks]: https://img.shields.io/github/forks/nfephp-org/sped-ibpt.svg?style=flat-square
[ico-issues]: https://img.shields.io/github/issues/nfephp-org/sped-ibpt.svg?style=flat-square

[ico-downloads]: https://img.shields.io/packagist/dt/nfephp-org/sped-ibpt.svg?style=flat-square
[ico-version]: https://img.shields.io/packagist/v/nfephp-org/sped-ibpt.svg?style=flat-square
[ico-license]: https://poser.pugx.org/nfephp-org/nfephp/license.svg?style=flat-square

[link-actions]: https://github.com/nfephp-org/sped-gtin/actions
[link-packagist]: https://packagist.org/packages/nfephp-org/sped-ibpt
[link-downloads]: https://packagist.org/packages/nfephp-org/sped-ibpt
[link-author]: https://github.com/nfephp-org
[link-issues]: https://github.com/nfephp-org/sped-ibpt/issues
[link-forks]: https://github.com/nfephp-org/sped-ibpt/network
[link-stars]: https://github.com/nfephp-org/sped-ibpt/stargazers


CHANGELOG.md
# Changelog

All Notable changes to `sped-ibpt` will be documented in this file.

Todas as atualizações a partir de 09/11/2016 devem observar os principios [Mantendo o CHANGELOG](http://keepachangelog.com/).

## 1.0.0 
Primeira release

## 1.0.1
-Inclusão de testes unitários

## 1.0.2
-Correções menores em README e comentários

## 2.0.0
- Alteração de API para HTTPS 
  NOTA: somente essa versão será aceita após 05 de agosto de 2018

tests/IbptTest.php
<?php

/**
 * This file belongs to the NFePHP project
 * php version 7.0 or higher
 *
 * @category  Library
 * @package   NFePHP\Ibpt\Rest
 * @author    Roberto L. Machado <linux.rlm@gmail.com>
 * @copyright 2016 NFePHP Copyright (c)
 * @license   https://opensource.org/licenses/MIT MIT
 * @link      http://github.com/nfephp-org/sped-ibpt
 */

namespace NFePHP\Ibpt\Tests;

use NFePHP\Ibpt\Ibpt;
use PHPUnit\Framework\TestCase;

/**
 * Unit Tests
 *
 * @category  Library
 * @package   NFePHP\Ibpt\Rest
 * @author    Roberto L. Machado <linux.rlm@gmail.com>
 * @copyright 2016 NFePHP Copyright (c)
 * @license   https://opensource.org/licenses/MIT MIT
 * @link      http://github.com/nfephp-org/sped-ibpt
 */
class IbptTest extends TestCase
{
    /**
     *
     * @var Ibpt
     */
    public $ibpt;
    /**
     *
     * @var object
     */
    public $dummyRest;
    /**
     *
     * @var array
     */
    public $expectedProd = [
        'Codigo' => '60063210',
        'UF' => 'SP',
        'EX' => 0,
        'Descricao' => 'Tecidos de malha de fibras sinteticas, tintos, de náilon ou de outras poliamidas',
        'Nacional' => 13.45,
        'Estadual' => 18,
        'Importado' => 36.08,
        'Municipal' => 0,
        'Tipo' => 0,
        'VigenciaInicio' => '01/04/2018',
        'VigenciaFim' => '30/06/2018',
        'Chave' => 'F3W1D7',
        'Versao' => '18.1.B',
        'Fonte' => 'IBPT/empresometro.com.br',
        'Valor' => 60,
        'ValorTributoNacional' => 8.07,
        'ValorTributoEstadual' => 10.8,
        'ValorTributoImportado' => 21.65,
        'ValorTributoMunicipal' => 0
    ];
    /**
     *
     * @var array
     */
    public $expectedServ = [
        'Codigo' => '0107',
        'UF' => 'SP',
        'Descricao' => 'Suporte técnico em informática, inclusive instalação, '
        . 'configuração e manutenção de programas de computação e bancos de dados.',
        'Tipo' => 2,
        'Nacional' => 13.45,
        'Estadual' => 0,
        'Municipal' => 2.7,
        'Importado' => 15.45,
        'VigenciaInicio' => '01/04/2018',
        'VigenciaFim' => '30/06/2018',
        'Chave' => 'F3W1D7',
        'Versao' => '18.1.B',
        'Fonte' => 'IBPT/empresometro.com.br',
        'Valor' => 500,
        'ValorTributoNacional' => 67.25,
        'ValorTributoEstadual' => 0,
        'ValorTributoImportado' => 77.25,
        'ValorTributoMunicipal' => 13.5
    ];


    /**
     * Set UP
     */
    public function setUp()
    {
        $this->dummyRest = $this->getMockBuilder('\NFePHP\Ibpt\RestInterface')
            ->setMethods(['pull'])
            ->getMock();
        $this->dummyRest->method('pull')->willReturn(json_encode($this->expectedProd));
        $this->ibpt = new Ibpt('12345678901234', 'a5354bf3890s0849', [], $this->dummyRest);
    }

    /**
     * Test check instanciate
     */
    public function testShouldInstantiate()
    {
        $this->assertInstanceOf(Ibpt::class, $this->ibpt);
    }

    /**
     * Test Product Taxes
     */
    public function testProductTaxes()
    {
        $expected = json_decode(json_encode($this->expectedProd));
        $resp = $this->ibpt->productTaxes(
            'SP',
            '60063210',
            0,
            'Tecidos de malha',
            'kg',
            60.00,
            'SEM GTIN'
        );
        $this->assertEquals($expected, $resp);
    }

    public function testServiceTaxes()
    {
        $this->dummyRest = $this->getMockBuilder('\NFePHP\Ibpt\RestInterface')
            ->setMethods(['pull'])
            ->getMock();

        $this->dummyRest->method('pull')->willReturn(json_encode($this->expectedServ));

        $this->ibpt = new Ibpt('12345678901234', 'a5354bf3890s0849', [], $this->dummyRest);

        $expected = json_decode(json_encode($this->expectedServ));
        $resp = $this->ibpt->serviceTaxes(
            'SP',
            '0107',
            'Suporte técnico em informática',
            'un',
            500.00
        );
        $this->assertEquals($expected, $resp);
    }
}


tests/HttpCodeTest.php
<?php

namespace NFePHP\Ibpt\Tests;

use NFePHP\Ibpt\HttpCode;

class HttpCodeTest extends \PHPUnit_Framework_TestCase
{
    public function testInfo()
    {
        $resp = HttpCode::info(100);
        $expected = "Informational";
        $this->assertEquals($expected, $resp['level']);
    }
}


src/Ibpt.php
<?php

/**
 * This file belongs to the NFePHP project
 * php version 7.0 or higher
 *
 * @category  Library
 * @package   NFePHP\Ibpt\Ibpt
 * @copyright 2016 NFePHP Copyright (c)
 * @license   https://opensource.org/licenses/MIT MIT
 * @author    Roberto L. Machado <linux.rlm@gmail.com>
 * @link      http://github.com/nfephp-org/sped-ibpt
 */

namespace NFePHP\Ibpt;

use NFePHP\Ibpt\RestInterface;
use NFePHP\Ibpt\Rest;

/**
 * Class to get taxes informations for consumers from IBPT
 *
 * @category  Library
 * @package   NFePHP\Ibpt\Ibpt
 * @author    Roberto L. Machado <linux.rlm@gmail.com>
 * @copyright 2016 NFePHP Copyright (c)
 * @license   https://opensource.org/licenses/MIT MIT
 * @link      http://github.com/nfephp-org/sped-ibpt
 */
class Ibpt
{
    /**
     * URL
     *
     * @var string
     */
    protected $url = 'https://apidoni.ibpt.org.br/api/v1/';
    /**
     * CNPJ do autorizado a fazer a cunsulta no IBPT
     *
     * @var string
     */
    private $cnpj;
    /**
     * Token do IBPT
     *
     * @var string
     */
    private $token;
    /**
     * Classe Rest Client
     *
     * @var object|null
     */
    protected $rest;

    /**
     * Constructor
     *
     * @param string $cnpj  cnpj do autorizado
     * @param string $token token do autorizado
     * @param array $proxy dados para proxy
     * @param object|null $rest interface
     */
    public function __construct(
        $cnpj,
        $token,
        $proxy = [],
        $rest = null
    ) {
        $this->cnpj = $cnpj;
        $this->token = $token;
        $this->rest = $rest;
        if (empty($rest)) {
            $this->rest = new Rest($proxy);
        }
    }

    /**
     * Get informations about produtcts taxes from IBPT restful service
     *
     * @param string $uf sigla do estado
     * @param string $ncm numero do ncm
     * @param int    $extarif numero do ex tarifario
     * @param string $descricao descrição
     * @param string $unidadeMedida unidade de medida
     * @param number $valor valor
     * @param string $gtin codigo gtin
     * @param string $codigoInterno codigo interno
     *
     * @return \stdClass
     *
     * @throws \Exception
     */
    public function productTaxes(
        $uf,
        $ncm,
        $extarif,
        $descricao,
        $unidadeMedida,
        $valor,
        $gtin,
        $codigoInterno = ''
    ) {
        $uri = $this->url
            . "produtos?token=" . $this->token
            . "&cnpj=" . $this->cnpj
            . "&codigo=$ncm"
            . "&uf=$uf"
            . "&ex=$extarif";

          $uri .= !empty($codigoInterno) ? "&codigoInterno=" . rawurlencode($codigoInterno) : "";
          $uri .= '&descricao=' . rawurlencode($descricao);
          $uri .= '&unidadeMedida=' . rawurlencode($unidadeMedida);
          $uri .= "&valor=$valor";
          $uri .= '&gtin=' . rawurlencode($gtin);
        return json_decode($this->rest->pull($uri));
    }

    /**
     * Get informations about services taxes from IBPT restful service
     *
     * @param string $uf uf
     * @param string $codigo codigo do serviço
     * @param string $descricao descricao
     * @param string $unidadeMedida unidade de medida
     * @param number $valor valor
     *
     * @return \stdClass
     */
    public function serviceTaxes(
        $uf,
        $codigo,
        $descricao,
        $unidadeMedida,
        $valor
    ) {
        $uri = $this->url
            . "servicos?token="
            . $this->token
            . "&cnpj="
            . $this->cnpj
            . "&codigo=$codigo"
            . "&uf=$uf";
        $uri .= "&descricao=" . rawurlencode($descricao);
        $uri .= "&unidadeMedida=" . rawurlencode($unidadeMedida);
        $uri .= "&valor=$valor";
        return json_decode($this->rest->pull($uri));
    }
}


src/Rest.php
<?php

/**
 * This file belongs to the NFePHP project
 * php version 7.0 or higher
 *
 * @category  Library
 * @package   NFePHP\Ibpt\Rest
 * @author    Roberto L. Machado <linux.rlm@gmail.com>
 * @copyright 2016 NFePHP Copyright (c)
 * @license   https://opensource.org/licenses/MIT MIT
 * @link      http://github.com/nfephp-org/sped-ibpt
 */

namespace NFePHP\Ibpt;

use NFePHP\Ibpt\RestInterface;

/**
 * Class to conect to restful services
 *
 * @category  Library
 * @package   NFePHP\Ibpt\Rest
 * @author    Roberto L. Machado <linux.rlm@gmail.com>
 * @copyright 2016 NFePHP Copyright (c)
 * @license   https://opensource.org/licenses/MIT MIT
 * @link      http://github.com/nfephp-org/sped-ibpt
 */
class Rest implements RestInterface
{

    /**
     * Parametros do proxy
     *
     * @var array
     */
    protected $proxy = [];

    /**
     * Constructor
     *
     * @param array $proxy Parameter for proxy ['IP','PORT','USER','PASS']
     */
    public function __construct($proxy = [])
    {
        if (!empty($proxy)) {
            $this->proxy = $proxy;
        }
    }

    /**
     * Pull data form IBPT Restful service to obtain taxes values
     *
     * @param string $uri
     *
     * @return string
     *
     * @throws \Exception
     */
    public function pull($uri)
    {
        $oCurl = curl_init($uri);
        if (!empty($this->proxy)) {
            $this->setProxy($oCurl, $this->proxy);
        }
        curl_setopt($oCurl, CURLOPT_RETURNTRANSFER, 1);
        curl_setopt($oCurl, CURLOPT_TIMEOUT, 10);
        curl_setopt($oCurl, CURLOPT_CONNECTTIMEOUT, 30);
        curl_setopt($oCurl, CURLOPT_SSL_VERIFYHOST, 0);
        curl_setopt($oCurl, CURLOPT_SSL_VERIFYPEER, 0);
        curl_setopt($oCurl, CURLOPT_SSLVERSION, 0);
        $response = curl_exec($oCurl);
        $soaperror = curl_error($oCurl);
        $soaperror_code = curl_errno($oCurl);
        $httpcode = curl_getinfo($oCurl, CURLINFO_HTTP_CODE);
        $ret = [];
        $ret['error'] = !empty(curl_error($oCurl)) ? curl_error($oCurl) : 'SUCESSO';
        $ret['response'] = $response;
        $ret['httpcode'] = $httpcode;
        curl_close($oCurl);
        if (intval($soaperror_code) != 0) {
            throw new \Exception("Erro cURL [{$soaperror_code}] {$soaperror}");
        }
        if ($httpcode != 200) {
            $response = json_encode($ret);
        }
        return $response;
    }

    /**
     * Set proxy parameters
     *
     * @param resource $oCurl
     * @param array $proxy
     *
     * @return resource
     */
    protected function setProxy(&$oCurl, $proxy)
    {
        curl_setopt($oCurl, CURLOPT_HTTPPROXYTUNNEL, 1);
        curl_setopt($oCurl, CURLOPT_PROXYTYPE, CURLPROXY_HTTP);
        curl_setopt($oCurl, CURLOPT_PROXY, $proxy['IP'] . ':' . $proxy['PORT']);
        if ($proxy['PASS'] != '') {
            curl_setopt($oCurl, CURLOPT_PROXYUSERPWD, $proxy['USER'] . ':' . $proxy['PASS']);
            curl_setopt($oCurl, CURLOPT_PROXYAUTH, CURLAUTH_BASIC);
        }
        return $oCurl;
    }
}


src/HttpCode.php
<?php

namespace NFePHP\Ibpt;

/**
 * HttpCode return a description os HTTP Codes returned from server
 * useful to help identify the cause of the communication problem,
 * either with a SOAP server or a server RESTFUL
 * The codes and their descriptions are stored in a json file in the same folder
 *
 * @category  NFePHP
 * @package   NFePHP\Ibpt\HttpCode
 * @copyright NFePHP Copyright (c) 2016
 * @license   http://www.gnu.org/licenses/lgpl.txt LGPLv3+
 * @license   https://opensource.org/licenses/MIT MIT
 * @license   http://www.gnu.org/licenses/gpl.txt GPLv3+
 * @author    Roberto L. Machado <linux.rlm at gmail dot com>
 * @link      http://github.com/nfephp-org/sped-common for the canonical source repository
 */

class HttpCode
{
    /**
     * Return informations about http code
     * @param int $code
     * @return array
     */
    public static function info($code)
    {
        $codes = (array) json_decode(file_get_contents(__DIR__.'/httpcodes.json'), true);
        if (array_key_exists($code, $codes)) {
            return $codes[$code];
        }
        return ['level' => 'Desconhecido', 'description' => 'Desconhecido', 'means' => 'Desconhecido'];
    }
}


examples/testeIBPT.php
<?php

error_reporting(E_ALL);
ini_set('display_errors', 'On');
require_once '../bootstrap.php';

use NFePHP\Ibpt\Ibpt;

$token = "<Aqui voce coloca seu token do IBPT>";
$cnpj = "<seu CNPJ>";
$ncm = "60063100"; //coloque o NCM do produto
$uf = 'MG'; //coloque o estado que deseja saber os dados
$extarif = 0; //indique o numero da exceção tarifaria, se existir ou deixe como zero

//instancia a classe Ibpt
$ibpt = new Ibpt($cnpj, $token);

//executa a consulta ao IBPT, o retorno é em stdClass
$resp = $ibpt->productTaxes($uf, $ncm, $extarif);

//caso não haja um retorno o erro e outras informações serão retornadas no JSON
echo "<pre>";
print_r($resp); //aqui mostra o retorno em um stdClass
echo "</pre>";

/*
 Em caso de SUCESSO irá retornar:

    stdClass Object
    (
        [Codigo] => 60063110
        [UF] => MG
        [EX] => 0
        [Descricao] => Tecidos de malha de fibras sinteticas, crus ou branqueados, de náilon ou de outras poliamidas
        [Nacional] => 13.45
        [Estadual] => 18
        [Importado] => 19.72
        [Municipal] => 0
        [Tipo] => 0
        [VigenciaInicio] => 26/10/2016
        [VigenciaFim] => 31/12/2016
        [Chave] => E13pH1
        [Versao] => 16.2.B
        [Fonte] => IBPT
    )

Em caso de não encontrar o produto pelo NCM, ou qualquer outro erro na comunicação, retornará:

   stdClass Object
   (
        [error] => SUCESSO
        [response] => "Produto não encontrado"
        [httpcode] => 404
    )
 */


examples/testeIBPTProduto.php
<?php

error_reporting(E_ALL);
ini_set('display_errors', 'On');
require_once '../bootstrap.php';

use NFePHP\Ibpt\Ibpt;

$token = "<coloque seu token>";
$cnpj = "<indique seu CNPJ sem formatação>";


$ncm = "60063210"; //coloque o NCM do produto
$uf = 'SP'; //coloque o estado que deseja saber os dados
$extarif = 0; //indique o numero da exceção tarifaria, se existir ou deixe como zero
$codigoInterno = '';
$descricao = 'Tecido';
$unidadeMedida = 'kg';
$valor = '60.00';
$gtin = 'SEM GTIN';


$ibpt = new Ibpt($cnpj, $token);

//executa a consulta ao IBPT, o retorno é em JSON
$resp = $ibpt->productTaxes(
    $uf,
    $ncm,
    $extarif,
    $descricao,
    $unidadeMedida,
    $valor,
    $gtin,
    $codigoInterno
);
//$resp = $ibpt->serviceTaxes($uf, $code);
//caso não haja um retorno o erro e outras informações serão retornadas no JSON
echo "<pre>";
print_r($resp); //aqui mostra o retorno em um stdClass
echo "</pre>";
